File: ERPEscolar.API/Models/Fiscal.cs
namespace ERPEscolar.API.Models;

/// <summary>
/// Factura electrónica (CFDI 4.0).
/// Almacena metadatos y referencias a XML/PDF/UUID.
/// </summary>
public class CFDI
{
    public int Id { get; set; }
    public int CargoId { get; set; }
    public string UUID { get; set; } = null!; // Folio fiscal del SAT
    public string Serie { get; set; } = null!; // Serie de la factura
    public string Folio { get; set; } = null!; // Folio de la factura
    public string RFC_Emisor { get; set; } = null!; // RFC de la escuela
    public string RFC_Receptor { get; set; } = null!; // RFC del tutor/alumno
    public string NombreReceptor { get; set; } = null!;
    public decimal Subtotal { get; set; }
    public decimal Descuento { get; set; } = 0;
    public decimal IVA { get; set; }
    public decimal Total { get; set; }
    public string Estado { get; set; } = "Borrador"; // "Borrador", "Timbrada", "Cancelada", "Error"
    public string? RazonCancelacion { get; set; }
    
    // Almacenamiento de documentos
    public string? RutaXML { get; set; } // Blob Storage path
    public string? RutaPDF { get; set; } // Blob Storage path
    public string? XMLConTimbrado { get; set; } // XML firmado y timbrado (completo)
    
    // Complemento Educativo (IEDU 1.0)
    public string? NivelEducativo { get; set; } // "Preescolar", "Primaria", "Secundaria", etc.
    public string? CURP_Alumno { get; set; }
    public string? ClaveCT { get; set; } // Clave de Centro de Trabajo
    
    // Auditoría fiscal
    public string? CadenaOriginal { get; set; } // Para validación
    public DateTime FechaTimbrado { get; set; } = DateTime.UtcNow;
    public DateTime? FechaCancelacion { get; set; }
    public string? ErrorTimbrado { get; set; }
    public int ReintentosTimbrado { get; set; } = 0;
    
    public Cargo Cargo { get; set; } = null!;
    public ICollection<BitacoraFiscal> Bitacoras { get; set; } = [];
}

/// <summary>
/// Bitácora fiscal (auditoría de cambios, intentos de timbrado, etc.).
/// Requerimiento legal: trazabilidad completa de operaciones fiscales.
/// </summary>
public class BitacoraFiscal
{
    public int Id { get; set; }
    public int? CFDIId { get; set; }
    public int? CargoId { get; set; }
    public string Evento { get; set; } = null!; // "Creación", "Intento_Timbrado", "Timbrado_Exitoso", "Cancelación", "Error"
    public string Descripcion { get; set; } = null!;
    public string? DetalleError { get; set; }
    public string? UsuarioId { get; set; }
    public string? IPOrigen { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
    
    public CFDI? CFDI { get; set; }
    public Cargo? Cargo { get; set; }
}

/// <summary>
/// Configuración de CFDI/Timbrado (credenciales de PAC, URLs, etc.).
/// </summary>
public class ConfiguracionCFDI
{
    public int Id { get; set; }
    public int SchoolId { get; set; }
    public string ProvedorTimbrado { get; set; } = null!; // "FINKOK", "QUADRUM", etc.
    public string UrlWSTimbrado { get; set; } = null!;
    public string UrlWSCancelacion { get; set; } = null!;
    public string Usuario_PAC { get; set; } = null!;
    public string Contrasena_PAC { get; set; } = null!; // Encriptada
    public string NombreComercial { get; set; } = null!; // Para PDF
    public string LogoUrlRuta { get; set; } = null!; // Blob path
    public int DiasCadenaOriginal { get; set; } = 30; // Días a mantener la cadena original en BD
    public int ReintentosMaximos { get; set; } = 3;
    public int TiempoEsperaReintentos { get; set; } = 300; // segundos
    public bool Activa { get; set; } = true;
    public DateTime FechaCreacion { get; set; } = DateTime.UtcNow;
    public DateTime? FechaActualizacion { get; set; }

    public School School { get; set; } = null!;
}

/// <summary>
/// Complemento educativo (IEDU) - Datos específicos para facturación educativa.
/// </summary>
public class ComplementoEducativo
{
    public int Id { get; set; }
    public int CFDIId { get; set; }
    public string NivelEducativo { get; set; } = null!; // "Preescolar", "Primaria", "Secundaria", "Bachillerato", "Profesional", "Postgrado"
    public string CURP { get; set; } = null!;
    public string ClaveCT { get; set; } = null!; // Clave de Centro de Trabajo (SEP)
    public string? NoExamenGeneral { get; set; }
    public DateTime FechaExpedicion { get; set; }
    public bool ValidadoSAT { get; set; } = false;

    public CFDI CFDI { get; set; } = null!;
}

File: erp-web/src/types/fiscal.ts

import type { Cargo } from './finanzas';

/**
 * Representa los datos básicos de un CFDI (Factura).
 */
export interface CFDI {
  id: number;
  cargoId: number;
  uuid: string;
  serie: string;
  folio: string;
  rfcEmisor: string;
  rfcReceptor: string;
  nombreReceptor: string;
  subtotal: number;
  descuento: number;
  iva: number;
  total: number;
  estado: 'Borrador' | 'Timbrada' | 'Cancelada' | 'Error';
  nivelEducativo?: string;
  curpAlumno?: string;
  claveCt?: string;
  fechaTimbrado?: string;
  fechaCancelacion?: string;
  razonCancelacion?: string;
  errorTimbrado?: string;
}

/**
 * Representa una entrada en la bitácora de eventos de un CFDI.
 */
export interface BitacoraFiscal {
  id: number;
  evento: string;
  descripcion: string;
  detalleError?: string;
  usuarioId: string;
  timestamp: string;
}

/**
 * Representa un CFDI con todos sus datos relacionados (cargo, bitácora, etc.).
 */
export interface CFDIFull extends CFDI {
  colegiatura: Cargo | null;
  cadenaOriginal?: string;
  xmlConTimbrado?: string;
  bitacoras: BitacoraFiscal[];
}

/**
 * DTO para crear un nuevo CFDI.
 */
export interface CreateCFDIDto {
  cargoId: number;
  rfcReceptor: string;
  nombreReceptor: string;
  serie?: string;
  folio?: string;
  nivelEducativo?: string;
  curpAlumno?: string;
  claveCt?: string;
}

/**
 * DTO para actualizar el estado de un CFDI (ej. para cancelar).
 */
export interface UpdateCFDIDto {
  estado: 'Borrador' | 'Timbrada' | 'Cancelada' | 'Error';
  razonCancelacion?: string;
  errorTimbrado?: string;
}

/**
 * DTO para solicitar el timbrado de un CFDI.
 */
export interface TimbrarCFDIDto {
  cfdiId: number;
  forzarTimbrado?: boolean;
}

/**
 * DTO para solicitar la cancelación de un CFDI.
 */
export interface CancelarCFDIDto {
  cfdiId: number;
  razonCancelacion: string;
}

/**
 * Representa la respuesta del servicio de timbrado/cancelación.
 */
export interface TimbradoResponse {
  exitoso: boolean;
  uuid?: string;
  mensaje: string;
  fechaTimbrado?: string;
  error?: string;
}
